### DISCUSSION TITLE
Is it a bug ?: After submit formState isValid is true when there are form errors

### DISCUSSION THREAD

**Discussion by ZaruhiTshitoyan** [react-hook-form/react-hook-form#8630] [Q&A]
After submitting form I get formState errors, but when making any field dirty it makes isValid to become true and my submit button is not disabled because of that. So basically, I have errors showing in the form but submit button is enabled ;/ I am checking !(isDIrty && isValid) for enabling or disabling my submit button.

Is it a bug? Any ideas? thank you ; )

**Comment by Moshyfawn** [comment]
You'll need to provide a repro sandbox. Can't reproduce

**Reply by ZaruhiTshitoyan** [reply]
https://codesandbox.io/s/exciting-joliot-81h17i

Here is the sandbox, is it pretty much similar to my code. The only difference is that I got error from the API telling that email already exists and I call setError from useEffect to show in the form, here I simply call setError after submit just to keep it simple. Then when I change first name or last name(just making them dirty) and not fixing the email error, still my submit button is enabled.

My idea: here for disabling button I can additionally check my errors object like this ===> !(isDirty && isValid && !Object.keys(errors).length) and the button is of course disabled. but it's additional check we don't want to.

Thank you for your time !

**Reply by bluebill1049** [reply]
https://react-hook-form.com/api/useform/seterror

> This method does not affect isValid formState, asisValid will always be derived by the validation result from your input registration rules or schema result.

**Reply by Moshyfawn** [reply]
I thought it's that, @bluebill1049, but I wasn't sure about it from the way the behavior was described

**Reply by IanVS** [reply]
Is there a way to detect the form state validity _including_ errors that were set with `setError()` (for example from an api response)?  Do I just need to look at `Object.keys(errors)`, maybe?

**Reply by bluebill1049** [reply]
For server response errors, you may find this useful: https://github.com/react-hook-form/react-hook-form/releases/tag/v7.43.0-next.0

**Reply by IanVS** [reply]
Thanks, yes I'm excited for that!  But my server also can return field-level errors, which I map back to the form and use `setError` to show.

**Reply by bluebill1049** [reply]
oh, interesting. Yea you will have to check the errors object for now, because `isValid` state is reflected on the field validation or schema validation result unless we start to take into consideration manual set errors internally.

This is pretty much the logic

```tsx
const isValid = isScahme ? runSchemaValidation() : runBuldinValiation();
```

To achieve what you describe above, we will have to switch to the following

```tsx
let isValid = isScahme ? runSchemaValidation() : runBuldinValiation();
isValid = isValid && isEmptyObject(formState.errors)
```

I think having a single source of truth to reflect `isValid` is preferred but happy to hear others' suggestions on this.

**Reply by IanVS** [reply]
Personally it seems to be like schema validation is only half of the story, and that manually-set errors should make the form invalid as well.  But I don't feel strongly about it, and I can use the check for `formState.errors` when I need it.  It just feels unexpected and like a possible footgun.  Maybe the docs should suggest checking for errors manually this way, if more than just the schema needs to be checked?

**Reply by bluebill1049** [reply]
I'm sorry for not getting back to you sooner, but I'm happy to discuss this issue more. maybe @Moshyfawn you want to give your opinion as well. As of today, our valid and dirty states all have a single source of truth, eg with isValid derived by schema/build validation. This creates confusion when the user set an error without setting the form to invalid, the problem I can see is when the user interacted with the form again and the validation re-run valid will be switching back to true unless we keep a set of all users set error and then the user will have to manually clear them (to set form back to valid). I am not 100% sure about going down that path or keep it as today and updating the doc.

**Reply by Moshyfawn** [reply]
> I'm sorry for not getting back to you sooner, but I'm happy to discuss this issue more. maybe @Moshyfawn you want to give your opinion as well. As of today, our valid and dirty states all have a single source of truth, eg with isValid derived by schema/build validation. This creates confusion when the user set an error without setting the form to invalid, the problem I can see is when the user interacted with the form again and the validation re-run valid will be switching back to true unless we keep a set of all users set error and then the user will have to manually clear them (to set form back to valid). I am not 100% sure about going down that path or keep it as today and updating the doc.

For sure, let me catch up with what's going on in this thread tomorrow 👌

**Reply by Moshyfawn** [reply]
So, currently, calling `setError` sets `isValid` to `false` until the next render, if all schema / built-in validation returns success. Is this the behaviour you want?

**Reply by IanVS** [reply]
Oh wow, thanks @Moshyfawn!  I did some investigating and indeed you're right.  It looked like it wasn't happening because in my form I have something that triggers a `elementRef.current.focus()` if the form is not valid after a submission, and it seems like it's that focus call which resets `isValid` to `true`.  Would you consider that to be a possible bug?  I can put together a reproduction if that would be helpful.

Overall, setting `isValid` to `false` after a `setError` until validation returns success does feel like the right thing to me.  Maybe the docs can be updated a bit to mention this?

**Reply by Moshyfawn** [reply]
> ...it seems like it's that focus call which resets `isValid` to `true`. Would you consider that to be a possible bug? I can put together a reproduction if that would be helpful.


A focus call generally shouldn't trigger a re-render as it doesn't change any React state. I tried focusing an input element `onSubmit` after setting an error via `setError` method, but I couldn't reproduce the behaviour you're describing. So, yea, you might have to put together a reproduction 😊

As a side note, RHF has a built-in API to focus the input you're attaching the error to:
```js
setError("firstName", { message: "Error Message" }, { shouldFocus: true });
```

**Reply by Moshyfawn** [reply]
> Overall, setting `isValid` to `false` after a `setError` until validation returns success does feel like the right thing to me. Maybe the docs can be updated a bit to mention this?

You're proposing that an error set via the `setError` method persists until.. the form is submitted again and the server returns success, correct? _Sorry, it's Monday evening, my brain processes information slower 🤭_

At this time, at least, RHF is not aware of the context the `setError` was called in _(an `onSubmit` callback, is it server response related etc.)_. `setError` is a generic form method to set a custom client form error and I don't think it can "become aware" of the server response status. That would be a different API, perhaps, we can consider as it won't tie into the current RHF design.

I do see your point, though, @IanVS: it does make sense to explore server / response form state.

Would you be able to expand on what you'd wanna see updated in the docs? You think they should mention / reflect something.

**Reply by IanVS** [reply]
OK I figured out what was going on, it wasn't the focus itself causing problems, it was blurring the field with an error.  That triggered the local validation to run, which reset the form to `isValid` because the local validation was fine.  Even though the `errors` aren't updated until the field is changed.

This started to happen in 7.39.0.  I confirmed it does not occur in 7.38.0, using codesandbox.  Here is my reproduction: https://codesandbox.io/s/react-hook-form-valid-after-blur-dx5xbi

**Reply by IanVS** [reply]
@Moshyfawn what do you think?  Is the behavior shown in the reproduction a bug, or intended?

**Reply by bluebill1049** [reply]
@IanVS could you try the latest patch? I just made a fix this morning, hopefully it's the same root cause.

**Reply by IanVS** [reply]
@bluebill1049 thanks.  It seems like the difference in 7.43.2 is that the field is not automatically focused as it was since 7.38.0.  However, focusing, and then blurring the field without changing the value still causes the `isValid` to become `true`.  

7.38.0 seems to work "correctly" (i.e. how I would expect), and 7.39+ does not, and the form becomes valid after simply focusing and blurring an input.

**Reply by bluebill1049** [reply]
I believe blurring the input will trigger the validation which updates the `isValid`, wondering what's the concern here in terms of UX. We basically re-validate the form validity with the user action.

**Reply by IanVS** [reply]
If the field has an error that was set by `setError`, ideally it wouldn't attempt revalidation until the value changes. After all, the error doesn't reset until there is a change.  And it doesn't make sense to me to have a form with errors and `isValid===false` that changes to `isValid===true` and showing the same errors after just moving the focus.

**Reply by bluebill1049** [reply]
> If the field has an error that was set by setError, ideally it wouldn't attempt revalidation until the value changes.

Yes, I agree, but at the same time. The form can't guarantee other fields have been impacted, so with `isValid` it's safer for us (library level) to re-run the validation for the entire form. Happy to hear everyone else's thoughts on this as well and work towards a common goal.
